server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java

```diff
@@ -145,6 +134,31 @@ public void addToFlushQueue(ShuffleDataFlushEvent event) {
     }
   }
 
+  public void processEvents(boolean endless) {
```

Comment:
Maybe we can create another method `eventLoop()`:

```java
  private void startEventProcesser() {
    // the thread for flush data
    Thread processEventThread = new Thread(this::eventLoop);
    processEventThread.setName("ProcessEventThread");
    processEventThread.setDaemon(true);
    processEventThread.start();
  }

  protected void eventLoop() {
    while (true) {
      processEvents();
    }
  }

  public void processEvents() {
    while (!flushQueue.isEmpty()) {
      try {
        ShuffleDataFlushEvent event = flushQueue.take();
        threadPoolExecutor.execute(() -> {
          processEvent(event);
        });
      } catch (Exception e) {
        LOG.error("Exception happened when process event.", e);
      }
    }
  }
```

Then override it in `TestShuffleFlushManager`:
```java
  @Override
  protected void eventLoop() {
    // do nothing
  }
```

Reply:
```java
protected void eventLoop() {
    while (true) {
      processEvents();
    }
  }

```

The CPU maybe will full?

Reply:
```java
while (true) {
  while (!flushQueue.isEmpty()) {
  }
}
```

is same with

```java
while (true || !flushQueue.isEmpty()) {
}
```

Reply:
Got. But in this pr, if `endless=true`, `flushQueue.take()` will block the thread.

Reply:
I didn't mean to change the `flushQueue.take()`.

Reply:
For reference: https://github.com/kaijchen/incubator-uniffle/tree/issue_509_fork

Reply:
I mean `flushQueue.take()` will  block the thread until `flushQueue` is not empty and it will not cost too much cpu resources. But `flushQueue.take()` will not block the thread in the code you implement and  it will not cost a lot of cpu resources.

Reply:
OK, I got it. If the flushQueue is empty, `processEvents()` will be called again and again.

Reply:
> For reference: https://github.com/kaijchen/incubator-uniffle/tree/issue_509_fork

We need overwrite `startEventProcessor` in `TestShuffleFlushManager`, Because if we start EventProcesser, events will be flushed automatically and this pr is used to prevent automatic flush.

Reply:
> We need overwrite `startEventProcessor` in `TestShuffleFlushManager`, Because if we start EventProcesser, events will be flushed automatically and this pr is used to prevent automatic flush.

I have fixed the infinite loop and squashed the commits, so you can see all changes in one commit.

Reply:
Update:
Actually I'm splitting `processEvents()` into `eventLoop()` and `flush()`.

```java
  protected void eventLoop() {
    while (true) {
      processNextEvent();
    }
  }

  @VisibleForTesting
  public void flush() {
    while (!flushQueue.isEmpty()) {
      processNextEvent();
    }
  }

  private void processNextEvent() {
    try {
      ShuffleDataFlushEvent event = flushQueue.take();
      threadPoolExecutor.execute(() -> {
        processEvent(event);
      });
    } catch (Exception e) {
      LOG.error("Exception happened when process event.", e);
    }
  }
```